### PR TITLE
Make `negative_bounds` internal & fix some of its issues

### PR DESCRIPTION
**Pull request by fmease**
r? compiler-errors

### CODE REVIEW

File: tests/ui/traits/negative-bounds/associated-constraints.rs [hunk 1]
```diff
@@ -17,4 +16,7 @@ fn test3<T: !Trait<Assoc: Send>>() {}
 fn test4<T>() where T: !Trait<Assoc: Send> {}
 //~^ ERROR associated type constraints not allowed on negative bounds
 
+fn test5<T>() where T: !Fn() -> i32 {}
```
Comment (fmease) [inline tests/ui/traits/negative-bounds/associated-constraints.rs:19]: This one used to ice with `no errors encountered even though `span_delayed_bug` issued` // `negative trait bounds should not have bindings`.

File: tests/ui/traits/negative-bounds/opaque-type-unsatisfied-bound.rs [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+    consume(produce()); //~ ERROR the trait bound `impl !Trait: Trait` is not satisfied
+}
+
+fn weird0() -> impl Sized + !Sized {}
+//~^ ERROR mismatched types
+//~| ERROR type mismatch resolving `() == impl !Sized`
+fn weird1() -> impl Sized + !Sized {}
+//~^ ERROR mismatched types
+//~| ERROR type mismatch resolving `() == impl !Sized`
+fn weird2() -> impl !Sized + Sized {}
```
Comment (fmease) [inline tests/ui/traits/negative-bounds/opaque-type-unsatisfied-bound.rs:21]: We elaborate `!Sized` to `!Sized + Sized` which I guess we should change?
Currently, I “unelaborate” any `Sized` bounds again even if we have a `!Sized` bound, too.

Comment (compiler-errors) [reply tests/ui/traits/negative-bounds/opaque-type-unsatisfied-bound.rs:21]: `!Sized` should opt out implicit sized bounds like `?Sized`

File: compiler/rustc_hir_analysis/src/astconv/bounds.rs [hunk 3]
```diff
@@ -71,15 +82,12 @@ impl<'tcx> dyn AstConv<'tcx> + '_ {
...
-        // If the above loop finished there was no `?Sized` bound; add implicitly sized if `Sized` is available.
-        if sized_def_id.is_none() {
-            // No lang item for `Sized`, so we can't add it as a bound.
-            return;
-        }
-        if seen_sized_unbound {
-            // There was in fact a `?Sized` bound, return without doing anything
-        } else {
-            // There was no `?Sized` bound; add implicitly sized if `Sized` is available.
+        if seen_sized_unbound || seen_negative_sized_bound {
```
Comment (fmease) [inline compiler/rustc_hir_analysis/src/astconv/bounds.rs:85]: I wonder if I should reject `!Sized + ?Sized`: (kinda) “redundant `?Sized` bound” // “`!Sized` already opts out of `Sized`”.

**Review by compiler-errors — APPROVED** [review]

**Pull request merged**